server/src/http/mod.rs

```diff
@@ -0,0 +1,124 @@
+                    let value = match value {
+                        query::Value::Null => serde_json::Value::Null,
+                        query::Value::Integer(i) => serde_json::Value::Number(Number::from(i)),
+                        query::Value::Real(x) => serde_json::Value::Number(
+                            Number::from_f64(x).ok_or(anyhow::anyhow!("invalid float value"))?,
+                        ),
+                        query::Value::Text(s) => serde_json::Value::String(s),
+                        query::Value::Blob(v) => {
+                            serde_json::Value::String(BASE64_STANDARD_NO_PAD.encode(v))
+                        }
+                    };
+
+                    map.insert(name.to_string(), value);
+                    Ok(map)
+                },
+            )?;
+
+        values.push(val);
+    }
+
+    let mut buffer = BytesMut::new().writer();
+    serde_json::to_writer(&mut buffer, &values)?;
+
+    Ok(buffer.into_inner().freeze())
+}
+
+async fn handle_request(
+    mut req: Request<Body>,
+    sender: mpsc::Sender<(oneshot::Sender<Result<QueryResponse, BoxError>>, Query)>,
+) -> anyhow::Result<Response<Body>> {
+    let bytes = to_bytes(req.body_mut()).await?;
+    let req: HttpQueryRequest = serde_json::from_slice(&bytes)?;
+    let (s, resp) = oneshot::channel();
+    let _ = sender
+        .send((s, Query::SimpleQuery(req.statements.join(";"), Vec::new())))
+        .await;
+
+    let result = resp.await?;
+    match result {
+        Ok(rows) => {
+            let json = query_response_to_json(rows)?;
+            Ok(Response::new(Body::from(json)))
+        }
+        Err(_) => todo!(),
+    }
+}
+
+pub async fn run_http<F>(addr: SocketAddr, db_factory: F) -> anyhow::Result<()>
+where
+    F: MakeService<(), Query> + Send + 'static,
+    F::Service: Load + Service<Query, Response = QueryResponse, Error = QueryError>,
+    <F::Service as Load>::Metric: std::fmt::Debug,
+    F::MakeError: Into<BoxError>,
+    F::Error: Into<BoxError>,
+    <F as MakeService<(), Query>>::Service: Send,
+    <F as MakeService<(), Query>>::Future: Send,
+    <<F as MakeService<(), Query>>::Service as Service<Query>>::Future: Send,
+{
+    tracing::info!("listening for HTTP requests on {addr}");
+
+    let (sender, mut receiver) = mpsc::channel(1024);
+    let server =
+        hyper::server::Server::bind(&addr).serve(make_service_fn(move |_: &AddrStream| {
+            let sender = sender.clone();
+            async move {
+                Ok::<_, Infallible>(service_fn(move |req| handle_request(req, sender.clone())))
+            }
+        }));
+
+    tokio::spawn(async move {
+        let mut pool = pool::Builder::new().build(db_factory, ());
+        while let Some((resp, query)) = receiver.recv().await {
+            if let Err(e) = poll_fn(|c| pool.poll_ready(c)).await {
+                tracing::error!("Connection pool error: {e}");
+                continue;
```

Comment:
Is it ok to drop the query here when the connection pool fails?
I'm not sure whether dropping a query and keeping the operation is the right choice here.
We have 2 other options:
1. fail the server
2. retry fetching the connection while keeping the same query

If dropping the query is what we really want then maybe we should at least include the query into the error message so that we know what was dropped?

Reply:
no it's not, but I fixed it upstream. Not RFR :)

Reply:
BTW, what's happening here is that the query is not executed, and an internal error will be returned to the user because the response channel was dropped

Reply:
Sounds good then. What's RFR btw?

Reply:
Ready for review